### PR TITLE
add router component and deep link support

### PR DESCRIPTION
**Pull request by ssseasonnn**
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## In the scope
add router component
add deeplink support

## Checklist
1. [x] Makes your submission pass tests?
2. [x] Have you lint your code locally before submission?
3. [ ] Have you updated the wiki of your change?

## Reference
- Card: https://trello.com/c/S26G7pKF/124-3-internal-router

### CODE REVIEW

**Review by jie-meng — APPROVED** [review]

File: app/proguard-rules.pro [hunk 1]
```diff
@@ -87,4 +87,14 @@
 -dontwarn okhttp3.internal.platform.**
 -dontwarn org.conscrypt.**
 -dontwarn org.bouncycastle.**
--dontwarn org.openjsse.**
\ No newline at end of file
+-dontwarn org.openjsse.**
+
+# Router
+-keep public class com.thoughtworks.ark.router.module.**
```
Comment (Tony---Zhang) [inline app/proguard-rules.pro:93]: 这个规则需要给sample也加一个？

File: app/src/androidTest/java/com/thoughtworks/ark/deeplink/DeeplinkManagerTest.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.net.URLEncoder
+
+@RunWith(AndroidJUnit4::class)
+class DeeplinkManagerTest {
+
+    @Test
+    fun should_return_empty_string_when_origin_scheme_is_empty() {
```
Comment (Tony---Zhang) [inline app/src/androidTest/java/com/thoughtworks/ark/deeplink/DeeplinkManagerTest.kt]: 我们的测试都是”驼峰“命名

File: app/src/main/java/com/thoughtworks/ark/MainActivity.kt [hunk 3]
```diff
@@ -1,30 +1,62 @@
...
+                        .group()
+                        .route(this)
+                }
+            }
+            true
+        }
+        navView.selectedItemId = selectedItemId
+    }
+
+    @SuppressLint("MissingSuperCall")
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/ark/MainActivity.kt]: 这个 可以删除吧？

File: app/src/main/java/com/thoughtworks/ark/deeplink/ui/DeeplinkActivity.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import com.thoughtworks.ark.deeplink.DeeplinkSchemeInterceptor
+import com.thoughtworks.ark.router.Router
+
+class DeeplinkActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        parseScheme(intent)
+    }
+
+    @SuppressLint("MissingSuperCall")
```
Comment (Tony---Zhang) [inline app/src/main/java/com/thoughtworks/ark/deeplink/ui/DeeplinkActivity.kt]: 这个也可以去掉吧？

File: build-logic/plugin/src/main/kotlin/com/thoughtworks/ark/Utils.kt [hunk 5]
```diff
@@ -13,6 +16,7 @@ import org.gradle.api.provider.Provider
 import org.gradle.kotlin.dsl.getByType
 import org.gradle.plugin.use.PluginDependency
 import org.jetbrains.kotlin.gradle.dsl.KotlinJvmOptions
+import java.util.*
```
Comment (Tony---Zhang) [inline build-logic/plugin/src/main/kotlin/com/thoughtworks/ark/Utils.kt:21]: 这个import *需要避免，我们可能这块缺乏detekt

**Review by Tony---Zhang — APPROVED** [review]

**Pull request merged**